src/material/table/table-data-source.ts

```diff
@@ -148,6 +148,14 @@ export class MatTableDataSource<T> extends DataSource<T> {
       let valueA = this.sortingDataAccessor(a, active);
       let valueB = this.sortingDataAccessor(b, active);
 
+      const valueAType = typeof valueA;
+      const valueBType = typeof valueB;
+      const sameType = valueAType === valueBType;
+      if (!sameType) {
+        valueA = valueAType === 'number' ? valueA.toString() : valueA;
+        valueB = valueBType === 'number' ? valueB.toString() : valueB;
```

Comment:
Using the ternary here seems redundant since it only applies to numbers. Maybe do something like this instead?

```
if (valueAType === 'number') {
  valueA += '';
}

if (valueBType === 'number') {
  valueB += '';
}
```

Also it might be a good idea to have a comment explaining the reasoning behind why they get stringified.